google-cloud-storage/src/main/java/org/jclouds/googlecloudstorage/domain/templates/ComposeObjectTemplate.java

```diff
@@ -14,25 +14,106 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-
 package org.jclouds.googlecloudstorage.domain.templates;
 
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import java.util.ArrayList;
+import java.util.Collection;
 import java.util.List;
 
 import org.jclouds.googlecloudstorage.domain.GoogleCloudStorageObject;
+import org.jclouds.javax.annotation.Nullable;
+import org.jclouds.json.SerializedNames;
 
 import com.google.auto.value.AutoValue;
+import com.google.common.collect.ImmutableList;
 
 @AutoValue
 public abstract class ComposeObjectTemplate {
 
-   private final String kind = "storage/composeRequest";
+   @AutoValue
+   public abstract static class SourceObject {
+
+      @AutoValue
+      public abstract static class ObjectPreconditions {
+         @Nullable public abstract long ifGenerationMatch();
+
+         @SerializedNames({"ifGenerationMatch"})
+         public static ObjectPreconditions create(long ifGenerationMatch){
+            return new AutoValue_ComposeObjectTemplate_SourceObject_ObjectPreconditions(ifGenerationMatch);
+         }
+
+         ObjectPreconditions(){
+         }
+      }
 
-   public abstract List<GoogleCloudStorageObject> sourceObjects();
+      public abstract String name();
+      @Nullable public abstract Long generation();
+      @Nullable public abstract ObjectPreconditions objectPreconditions();
+
+      public static SourceObject nameOnly(String name){
+         return create(name, null, null);
+      }
+
+      public static SourceObject createWithPrecondition(String name, Long generation, Long objectPreconditions){
+         return create(name, generation, ObjectPreconditions.create(objectPreconditions));
+      }
+
+      @SerializedNames({ "name", "generation", "objectPreconditions"})
+      public static SourceObject create(String name, @Nullable Long generation,
+            @Nullable ObjectPreconditions objectPreconditions) {
+         return new AutoValue_ComposeObjectTemplate_SourceObject(name, generation, objectPreconditions);
+      }
+
+      SourceObject(){
+      }
+   }
 
+   public abstract List<SourceObject> sourceObjects();
    public abstract ObjectTemplate destination();
 
-   public static ComposeObjectTemplate create(List<GoogleCloudStorageObject> sourceObjects, ObjectTemplate destination) {
+   @SerializedNames({"sourceObjects", "destination"})
+   public static ComposeObjectTemplate create(List<SourceObject> sourceObjects, ObjectTemplate destination) {
       return new AutoValue_ComposeObjectTemplate(sourceObjects, destination);
    }
+
+   ComposeObjectTemplate() {
+   }
+
+   public static class Builder {
+      private ImmutableList<SourceObject> sourceObjects;
+      private ObjectTemplate destination;
+
+      public Builder() {
```

Comment:
> I don't doubt there is a better way. Im not sure what you are envisioning, are you suggesting I put a  static `builder` method be in the ComposeObjectTemplate class and not ComposeObjectTemplate.Builder class and make the current `public Builder()` method package protected?

Correct; we should prefer a private constructor and public static `builder` method which calls it.